stream-chat-android-ui-components/src/test/kotlin/io/getstream/chat/android/ui/message/list/header/viewmodel/MessageListHeaderViewModelTest.kt

```diff
@@ -1,101 +1,146 @@
+private const val CHANNEL_NAME = "channel"
 
-/*
+@OptIn(ExperimentalCoroutinesApi::class)
 @ExtendWith(InstantTaskExecutorExtension::class)
 internal class MessageListHeaderViewModelTest {
 
-    private val chatDomain: ChatDomain = mock()
-    private val channelControllerResult: Result<ChannelController> = mock()
-    private val channelControllerCall = TestCall(channelControllerResult)
-    private val channelController: ChannelController = mock()
-
-    @BeforeEach
-    fun setup() {
-        whenever(chatDomain.user) doReturn MutableLiveData(CURRENT_USER)
-        whenever(chatDomain.watchChannel(CID, 0)) doReturn channelControllerCall
-        whenever(channelControllerResult.isSuccess) doReturn true
-        whenever(channelControllerResult.data()) doReturn channelController
+    private val members = createMembers { createMember(User(name = "user")) }
+
+    private val membersStateFlow = MutableStateFlow(createMembers())
+    private val channelDataStateFlow = MutableStateFlow(ChannelData(Channel()))
+
+    private val channelState: ChannelState = spy {
+        whenever(it.members).doAnswer { membersStateFlow }
+        whenever(it.channelData).doAnswer { channelDataStateFlow }
+        whenever(it.toChannel()).doAnswer { returnChannel() }
     }
 
-    @Test
-    fun `Should notify about a new channel`() {
-        val channel: Channel = mock()
-        whenever(channelController.offlineChannelData) doReturn MutableLiveData(mock())
-        whenever(channelController.toChannel()) doReturn channel
-        val channelHeaderViewModel = MessageListHeaderViewModel(CID, chatDomain = chatDomain)
-        val mockObserver: Observer<Channel> = spy()
-
-        channelHeaderViewModel.channelState.observeForever(mockObserver)
-        verify(mockObserver, times(2)).onChanged(eq(channel))
+    private val stateRegistry: StateRegistry = mock {
+        whenever(it.channel(any(), any())) doReturn channelState
+    }
+    private val chatClient: ChatClient = mock {
+        whenever(it.queryChannel(any(), any(), any())) doReturn Channel().asCall()
+    }
+
+    init {
+        StateRegistry.instance = stateRegistry
     }
 
+    @OptIn(ExperimentalCoroutinesApi::class)
     @Test
-    fun `Should notify about new members`() {
-        val members = createMembers()
-        whenever(channelController.members) doReturn MutableLiveData(members)
-        val channelHeaderViewModel = MessageListHeaderViewModel(CID, chatDomain = chatDomain)
-        val mockObserver: Observer<List<Member>> = spy()
-
-        channelHeaderViewModel.members.observeForever(mockObserver)
-        verify(mockObserver).onChanged(eq(members))
+    fun `When channel members update should update channel`() = runBlockingTest {
+        // given
+        val messageListHeaderViewModel = MessageListHeaderViewModel(CID, chatClient = chatClient)
+        var result: Channel? = null
+        launch {
+            result = messageListHeaderViewModel._channel.take(2).toList().last()
```

Comment:
why do we start a coroutine here to change a local variable?

Reply:
All other approaches resulted in this particular thread being blocked or the tests erroring out due to an unfinished job.

I'm open to suggestions because this isn't pretty, but attempts to make it prettier ended up in the tests not completing.